internal/services/users-stats-service.go

```diff
@@ -0,0 +1,63 @@
+package services
+
+import (
+	"github.com/afmireski/garchop-api/internal/models"
+	"github.com/afmireski/garchop-api/internal/ports"
+
+	customErrors "github.com/afmireski/garchop-api/internal/errors"
+	myTypes "github.com/afmireski/garchop-api/internal/types"
+)
+
+type UsersStatsService struct {
+	repository   ports.UserStatsRepository
+	tiersService *TiersService
+}
+
+func NewUsersStatsService(repository ports.UserStatsRepository, tiersService *TiersService) *UsersStatsService {
+	return &UsersStatsService{
+		repository,
+		tiersService,
+	}
+}
+
+func (s *UsersStatsService) LevelUp(userId string, newTierId uint) *customErrors.InternalError {
+	_, err := s.repository.Update(userId, myTypes.AnyMap{"tier_id": newTierId}, myTypes.Where{})
+
+	if err != nil {
+		return customErrors.NewInternalError("a failure occurred during user level up", 500, []string{err.Error()})
+	}
+
+	return nil
+}
+
+func (s *UsersStatsService) GainExperience(userId string, currentTierId uint, currentExperience uint, items []models.ItemModel) *customErrors.InternalError {
+
+	gainedXp := s.calculateGainedXp(items)
+	nextTier, findNextTierErr := s.tiersService.FindNextTier(int(currentTierId))
+
+	if findNextTierErr != nil {
+		return customErrors.NewInternalError("a failure occurred when try to find the next tier", 500, []string{findNextTierErr.Error()})
+	}
+	experienceSum := currentExperience + gainedXp
+
+	_, err := s.repository.Update(userId, myTypes.AnyMap{"experience": experienceSum}, myTypes.Where{}); if err != nil {
+		return customErrors.NewInternalError("a failure occurred during user experience gain", 500, []string{err.Error()})
+	}
+
+	if nextTier != nil && experienceSum >= nextTier.MinimalExperience {
+		s.LevelUp(userId, nextTier.Id)
+	}
+
+	return nil
+}
```

Comment:
```suggestion
func (s *UsersStatsService) GainExperience(userId string, experienceAmount uint) *customErrors.InternalError {

        // Buscar user_stats a partir do argumento userId (que no caso não temos pronto, acho que essa issue está comigo)
	nextTier, findNextTierErr := s.tiersService.FindNextTier(userStats.TierId)

	if findNextTierErr != nil {
		return customErrors.NewInternalError("a failure occurred when try to find the next tier", 500, []string{findNextTierErr.Error()})
	}
	experienceSum := userStats.Experience + experienceAmount

	_, err := s.repository.Update(userId, myTypes.AnyMap{"experience": experienceSum}, myTypes.Where{}); if err != nil {
		return customErrors.NewInternalError("a failure occurred during user experience gain", 500, []string{err.Error()})
	}

	if nextTier != nil && experienceSum >= nextTier.MinimalExperience {
		s.LevelUp(userId, nextTier.Id)
	}

	return nil
}
```

Não sei se faz tanto sentido a sugestão.

A ideia é que caso futuramente a gente implemente outra(s) forma(s) de ganhar experiência, podemos reaproveitar o método. Por exemplo, ganhar experiência a partir de uma caixa misteriosa, etc.